experimental: remove gulp from appium package

boneskull opened the pull request:
- Replaced gulp with babel, mocha, and npm scripts
    - Some redundancy necessary to support running scripts from monorepo root or `packages/appium`
    - `build/lib` => `build` now that we do not compile test files and have moved scripts from `commands-yml` to `scripts`. This may be a breaking change
- Moved `commands-yml/*.js` to scripts under `packages/appium/scripts/`.
- Refactored all scripts to be CJS, so they do not need transpilation.  Alternatively, we could have renamed them all `*.mjs` and used ESM, but most were CJS already.  For now, ESLint will disallow use of `import` keyword in this script folder for `.js` files.
- Normalized test file names to align with common convention (some IDEs now recognize the files as test files and behave accordingly)
    - `*-e2e-specs.js` => `*.e2e.spec.js`
    - `*-specs.js` => `*.spec.js`
- Reorganized tests to be more easily executed on CLI (E2E tests in `e2e`; all else in `unit`). Moved a fixture or two
- Sorted `package.json`s
- Moved `zip`/`upload`/`zip-and-upload` scripts to root (see "problems" below)
- Added `@babel/cli` dev dep
- Updated `.eslintignore`, `.eslintrc`, etc.
- Mocha config: added a check to ensure `.only()` doesn't get committed, sneak in to CI and give false positives

## Problems

- XXX: The `upload` script does not work, because there is no gulpfile here.  Is the `github-upload` task something that should be a script?  Do we use it elsewhere?
- ~~The `dev` script (and `watch` script) in the monorepo root `package.json` do not work, since we'd need another program (e.g., `concurrently`) to run `gulp --watch` and `babel --watch` in parallel.  When gulp is gone, this will get fixed.  For now, launch `dev` and the new `dev:appium` in separate terminals~~. 
**UPDATE:** This now works properly.

boneskull commented:
Current dependencies on/for this PR:
* 2.0
  * **PR #16443** <a href="https://app.graphite.dev/github/pr/appium/appium/16443" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #16482** <a href="https://app.graphite.dev/github/pr/appium/appium/16482" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #16459** <a href="https://app.graphite.dev/github/pr/appium/appium/16459" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

boneskull commented:
@jlipps Looking for guidance on the "github upload" script.  Do we still need it?  If we don't, I'll move this out of draft.

jlipps commented on packages/appium/scripts/parse-yml-commands.js:
```diff
@@ -1,13 +1,92 @@
+};
+
+const validator = {
+  'name': {presence: true},
```
though i get doing it here for consistency might look nicer

boneskull replied:
this is a copy-paste job

jlipps reviewed (approved):
looks good, however it doesn't seem to remove gulp? maybe i misunderstood the point

boneskull commented on package.json:
```diff
@@ -24,44 +24,49 @@
   "author": "https://github.com/appium",
   "scripts": {
-    "build": "lerna exec gulp transpile",
+    "build": "lerna exec --parallel -- gulp transpile",
```
Good catch, I missed this.

boneskull commented on packages/appium/scripts/write-server-args-docs.js:
```diff
@@ -0,0 +1,59 @@
+    'server-args.md',
+  ),
+);
+let md = '# Appium server arguments\n\n';
```
yes, might as well.

boneskull commented:
I'm going to revert the change which outputs the contents of `lib` into `build`; instead, the result will be under `build/lib` as it was before.  This would otherwise be a breaking change to [the java client](https://github.com/appium/java-client/blob/9fa0a13e2f1684bd996963b1f6bd1a962c879d21/src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java#L76) and [the python client](https://github.com/appium/python-client/blob/895cde1aa674aaab2f958ae251de0daefd049c02/appium/webdriver/appium_service.py#L26).  This should probably change (because `build` will _only_ contain `lib`), but it doesn't have to happen now.

boneskull committed:
chore: remove gulp from appium package

This change removes `gulp` and replaces it with plain ol' `babel` and `mocha`.

## `appium`

- No gulp.
- `test/` has been re-organized.  All test files now end in the extension `.spec.js`, which is a recognized convention and understood by editors/IDEs.  The tests are further split into `unit` and `e2e` subdirs.  This makes working with both `babel` and `mocha` easier.
- The tests are _not_ run against the transpiled code; code is now transpiled on-the-fly via `@babel/register`. This only affects `appium`.
- `commands-yml/validator.js` moved to `scripts/parse-yml-commands.js`. It has been rewritten as a CJS module.  `commands-yml/` should not contain `.js` files.
- `test/setup.js` is a new test harness specific to Appium, which loads & configures `@babel/register`.  Eventually, the contents of this file can be moved into the root monorepo's `test/setup.js`, but the rest of the packages do not need `@babel/register`, so we avoid the overhead.

## `@appium/gulp-plugins`

- Unfortunately, due to the new `npm test` strategy (see below), we cannot use the `nyan` reporter.  Or rather, we _can_, but will not see the animation.  So the reporter is now set to default to `spec`.
- Modified tests so they didn't overwrite the actual `build` dir with transpiled fixtures, which is what was happening. This was causing failures depending on the transpile/test order.

## All Packages

- All packages now have their own `build`, `test`, `test:e2e` and `dev` (just build + watch mode) npm scripts. This makes running a complete build easier, since each package can now provide its own appropriate command.  This is _mostly_ just `gulp once` (where `gulp` was used before), except for `appium`, where `@babel/cli` is used directly.
- This is what happens when `npm test` is run from the monorepo (in order):
    1. All packages are built (in parallel) by running `npm run build` in each folder (via `lerna`). See `pretest`
    2. All code is linted directly via `eslint`.
    3. All _unit tests_ are run w/ `npm test` in each folder, having assumed transpilation already occurred (if necessary).
- `npm run e2e-test` runs the `npm run test:e2e` in each folder. No transpilation occurs! Careful.
- All scripts are expected to live in `packages/*/scripts`, and this directory enforces CJS via ESLint.  `postinstall.js` moved, as well as `check-npm-pack-files.js`. `generate-schema-declarations.mjs` moved from the root `scripts/` dir into `packages/appium/scripts/`; it's now CJS instead of ESM.
- I had profiled running various things with `lerna run --parallel`, but it turns out this is not faster.  I'm not sure why, but I think it may have something to do with streaming STDOUT/STDERR, or potentially the overhead of running child processes outweighs the time it takes to actually run tests and builds.
- The behavior of `lerna run` now mimicks `lerna exec`, which was "run stuff in serial, streaming the output". The default behavior of `lerna run` is "run stuff in parallel, buffering the output".  Buffered output sucks, and streaming the output from all tasks at once creates chaos.  To override this behavior, you'll see `--parallel --concurrency=8 --prefix` for tasks that can truly be parallelized, like transpilation.  The number 8 is the "maximum", and CI usually won't have more than 2 cores available anyway.
- When run in CI, Mocha runs with `--forbid-only`, which will fail if there's an `.only()` somewhere in the tests. Further, it forces color output.

## Etc.

- Updated Wallaby config, `.gitignore`, `.eslintignore`.
- Sorted `package.json` files
- Removed `watch` and `coverage` scripts which were broken anyway.
